# Percepción (PER): examen de prácticas del bloque 1, grupo 3CO21, turno 2, 25-4-2023, 18:15-19:00

Lee este cuaderno y realiza las actividades y ejercicios propuestos.

## Importación de librerías relevantes

Como sabes, sklearn facilita el aprendizaje y evaluación de clasificadores generativos y discriminativos en digits. 

Ejecuta el código siguiente para importar librerías relevantes. Es posible que tengas que añadir alguna importación.

In [6]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import multivariate_normal
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression

## Lectura del corpus digits

Ejecuta el código siguiente con **random_state igual a las tres últimas cifras de tu DNI/NIE:**

In [2]:
digits = load_digits()
X = digits.data
y = digits.target
X_train, X_test, y_train, y_test = train_test_split(
    digits.data, digits.target, test_size=0.75, shuffle=True, random_state=921)

# Naive Bayes Gaussiano

Ejecuta el código siguiente para estudiar la precisión de naive Bayes Gaussiano en función de var_smoothing:

In [5]:
for v in [1e-9, 1e-5, 1e-1, .2, .5, 1]:
    clf = GaussianNB(var_smoothing=v).fit(X_train, y_train)
    acc = accuracy_score(y_test, clf.predict(X_test))
    print('La precisión de {0!s} es {1:.1%}'.format(clf, acc))

La precisión de GaussianNB() es 85.1%
La precisión de GaussianNB(var_smoothing=1e-05) es 87.7%
La precisión de GaussianNB(var_smoothing=0.1) es 91.2%
La precisión de GaussianNB(var_smoothing=0.2) es 90.1%
La precisión de GaussianNB(var_smoothing=0.5) es 89.5%
La precisión de GaussianNB(var_smoothing=1) es 88.9%


# Regresión logística

Añade una celda de código para estudiar el comportamiento del clasificador discriminativo basado en regresión logística en función del "solver" empleado. Busca el solver que maximice el acierto sobre el conjunto de test. Asegúrate que el parámetro de número máximo de iteraciones es suficiente para dejar que el algoritmo converja adecuadamente.

In [14]:
for solver in ['lbfgs', 'liblinear', 'newton-cg', 'newton-cholesky', 'sag', 'saga']:
    clf = LogisticRegression(solver=solver, max_iter=1000000).fit(X_train, y_train)
    acc = accuracy_score(y_test, clf.predict(X_test))
    print('La precisión de {0!s} es {1:.1%}'.format(clf, acc))

La precisión de LogisticRegression(max_iter=1000000) es 93.9%
La precisión de LogisticRegression(max_iter=1000000, solver='liblinear') es 93.0%
La precisión de LogisticRegression(max_iter=1000000, solver='newton-cg') es 94.0%
La precisión de LogisticRegression(max_iter=1000000, solver='newton-cholesky') es 94.1%
La precisión de LogisticRegression(max_iter=1000000, solver='sag') es 93.7%
La precisión de LogisticRegression(max_iter=1000000, solver='saga') es 93.6%


En este caso el solver que da mejores resultados es el **newton-cholesky** con un $94.1\%$